File: cpp/45.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <set>
#include <string>
#include <unordered_map>
#include <climits>
#include <numeric>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

class Solution {
public:
  int jump(int A[], int n) {
    int cur, last, step;
    cur = last = step = 0;
    for(int i=0; i<n; i++){
      if(i > last){
        ++step;
        last = cur;
      }
      cur = max(cur, i+A[i]);
    }
    return step;
  }
};

int main(){
  auto sol = Solution();
}

File: python/45.py
class Solution(object):
    def jump(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) <= 1:
            return 0
        cnt, best, last = 0, 0, -1
        for i, n in enumerate(nums):
            cur = i + nums[i]
            if i <= last:
                # print(cnt-1, best)
                # in the shadow of last jump
                if cur > best:
                    best = cur
            else:
                # print(cnt, best)
                # in the shadow best jump
                if i + nums[i] > best:
                    last = best
                    best = i + nums[i]
                    cnt += 1
            if best >= len(nums)-1:
                return cnt
        return -1

s = Solution()
print(s.jump([1]))
print(s.jump([1, 2]))
